### PR TITLE
[WIP] Node image builds for 3.10

### PR DESCRIPTION
**Pull request by kwoodson**
First attempt at node image build with e2e tests.

### CODE REVIEW

**Comment by kwoodson** [comment]
Currently failing with az commands not running in containers.

File: ci-operator/templates/build-azure-image-310.yaml [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+        ./hack/create.sh ${INSTANCE_PREFIX}
+
+        # if completed pass generated data to other stage
+        cp -r /go/src/github.com/openshift/openshift-azure/_data /tmp/shared/
+
+        touch /tmp/shared/e2e_tests
+
+    - name: tagimage
+      image: docker.io/openshift/origin-ansible:v3.10.1
```
Comment (stevekuznetsov) [inline ci-operator/templates/build-azure-image-310.yaml]: We should be doing this with the tag specification in ci-operator for the 3.10 branch, not here

Comment (kwoodson) [reply ci-operator/templates/build-azure-image-310.yaml]: Agreed.  I'm not sure I understand the tag specification completely.  This made it very easy.

Comment (stevekuznetsov) [reply ci-operator/templates/build-azure-image-310.yaml]: Hopefully the config reference is useful https://github.com/openshift/ci-operator/blob/master/CONFIGURATION.md#tag_specification

File: ci-operator/templates/build-azure-image-310.yaml [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    activeDeadlineSeconds: 10800
+    terminationGracePeriodSeconds: 600
+    volumes:
+    - name: shared-tmp
+      emptyDir: {}
+    - name: shared-data
+      emptyDir: {}
+    - name: cluster-secrets-azure
+      secret:
+        secretName: azure
```
Comment (mjudeikis) [inline ci-operator/templates/build-azure-image-310.yaml]: s/azure/e2e-azure-secret/

File: ci-operator/templates/build-azure-image-310.yaml [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+        #!/bin/bash
+        # error handling and sync code
+        set -euo pipefail
+
+        trap '[ $? == 0 ] || touch /tmp/shared/exit' EXIT
+        trap 'kill $(jobs -p); exit 0' TERM
+
+        # We need to persist this variable so that we can clean it up when we exit
+        echo "RESOURCE_GROUP=${INSTANCE_PREFIX}-$(TZ=Etc/UTC date +%Y%m%d%H%M)" >>/tmp/shared/environment
+        echo "IMAGE_NAME=centos7-$OPENSHIFT_RELEASE-$(TZ=Etc/UTC date +%Y%m%d%H%M)" >>/tmp/shared/environment
```
Comment (mjudeikis) [inline ci-operator/templates/build-azure-image-310.yaml]: What about RHEL images?

File: ci-operator/templates/build-azure-image-310.yaml [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+        # Get resource group and image name
+        source /tmp/shared/environment
+        echo "RESOURCE_GROUP=${RESOURCE_GROUP}"
+        echo "IMAGE_NAME=${IMAGE_NAME}"
+        echo $(pwd)
```
Comment (mjudeikis) [inline ci-operator/templates/build-azure-image-310.yaml]: delete

File: ci-operator/templates/build-azure-image-310.yaml [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+        export DNS_DOMAIN=osadev.cloud
+        export DNS_RESOURCEGROUP=dns
+        export DEPLOY_VERSION=v${OPENSHIFT_RELEASE}
+        export RUN_SYNC_LOCAL=true
+        export DEPLOY_OS="centos7"
+        export IMAGE_RESOURCEGROUP=images
+
+        source /tmp/shared/environment
+        export IMAGE_RESOURCENAME=${IMAGE_NAME}
+        # TODO: for full e2e image needs to be built too
```
Comment (mjudeikis) [inline ci-operator/templates/build-azure-image-310.yaml]: remove

File: ci-operator/templates/build-azure-image-310.yaml [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+        cd /go/src/github.com/openshift/openshift-azure/
+
+        # added timeout as this seems to loop infinitely awaiting cluster deployment
+        timeout 30m ./hack/create.sh ${RESOURCE_GROUP}-e2e
```
Comment (mjudeikis) [inline ci-operator/templates/build-azure-image-310.yaml]: What do you mean loops indefinitely? It shouldn't. What do you see if we dont have timeout?

**Comment by mjudeikis** [comment]
You might want to update : 
```
master.json	
release-3.10.json
release-3.11.json
``` 
to have similar format `tag_specification` to point to release of origin. 

What do you want to do with prow config for this template? we would need still to add it. Separate PR?

File: ci-operator/templates/build-azure-image-310.yaml (hunk 3 above)
Comment (kwoodson) [reply ci-operator/templates/build-azure-image-310.yaml]: Today we have separate jobs for the rhel and centos.  How do you foresee us having a single template and passing in variables for the build job?  I can variable-ize the flavor of OS but how does that get passed into the job?

File: ci-operator/templates/build-azure-image-310.yaml (hunk 4 above)
Comment (kwoodson) [reply ci-operator/templates/build-azure-image-310.yaml]: done

File: ci-operator/templates/build-azure-image-310.yaml (hunk 6 above)
Comment (kwoodson) [reply ci-operator/templates/build-azure-image-310.yaml]: I experienced failures where the cluster creation would wait indefinitely even though there were failures in the deployment.  This occurred when I shared the same resource group with the node build and the cluster build.  This pod would never complete so I added a timeout.

File: ci-operator/templates/build-azure-image-310.yaml (hunk 3 above)
Comment (kwoodson) [reply ci-operator/templates/build-azure-image-310.yaml]: @kargakis, any opinion on how this would work?

File: ci-operator/templates/build-azure-image-310.yaml (hunk 2 above)
Comment (mjudeikis) [reply ci-operator/templates/build-azure-image-310.yaml]: This still needs changing :)

File: ci-operator/config/openshift/openshift-azure/master.json [hunk 7]
```diff
@@ -1,8 +1,10 @@
 {
   "tag_specification": {
     "cluster": "https://api.ci.openshift.org",
-    "namespace": "azure",
-    "name": "plugin-base"
+    "namespace": "openshift",
+    "name": "origin",
```
Comment (kargakis) [inline ci-operator/config/openshift/openshift-azure/master.json]: Why this change?

**Comment by kargakis** [comment]
Please remove the branched jobs, they don't apply to us (at least yet)

File: ci-operator/config/openshift/openshift-azure/master.json (hunk 7 above)
Comment (mjudeikis) [reply ci-operator/config/openshift/openshift-azure/master.json]: This is thing we didn't do, but we suppose to. This tags all particular release images into our ci-namespace for us to use. If we need ansible,origin, or any other image from another team - this is how we get them so we dont need to build it ourselves.

File: ci-operator/templates/build-azure-image-310.yaml (hunk 2 above)
Comment (kwoodson) [reply ci-operator/templates/build-azure-image-310.yaml]: Yeah, I was still testing it locally.  I thought this was necessary for local tests.

**Comment by kwoodson** [comment]
@kargakis 
> Please remove the branched jobs, they don't apply to us (at least yet)

What did you mean by this?  The `release-*` files?  If that's the case, how do we get 3.10 builds since we need to specify a `origin-v3.10` in the release-3.10.yml file under `tag_specification`?

On a side note, how do we get 3.10 builds for centos and rhel without hard coding them into the template?  I put in a variable but I'm unsure how that variable gets populated for the actual run of the job.

**Comment by kargakis** [comment]
Your ci-operator jobs should be jobs we run in openshift-ansible and also periodically, right? The openshift-azure and azure-misc release jobs are not used.

**Comment by mjudeikis** [comment]
> Your ci-operator jobs should be jobs we run in openshift-ansible and also periodically, right? The openshift-azure and azure-misc release jobs are not used. 

we run then in openshift-azure but we consume artifacts from `openshift-ansible` runs. This way we dont need to rebuild ansible image each time.

**Comment by kwoodson** [comment]
```
2018/09/05 10:31:01 Running pod dev
2018/09/05 10:56:12 Container buildimage in pod dev completed successfully
2018/09/05 11:13:30 Container buildcluster in pod dev completed successfully
2018/09/05 11:14:22 Container tagimage in pod dev completed successfully
2018/09/05 11:14:59 Container teardown in pod dev completed successfully
2018/09/05 11:14:59 Pod dev succeeded after 43m59s
2018/09/05 11:14:59 Ran for 44m4s
```

**Comment by kargakis** [comment]
> we run then in openshift-azure but we consume artifacts from openshift-ansible runs. This way we dont need to rebuild ansible image each time.

Not sure I follow. What's the relation between openshift-azure and the node image? What's the overhead of building the ansible docker image?

**Comment by mjudeikis** [comment]
Easier to show than to tell :) tomorrow will show this.

**Comment by kargakis** [comment]
> The node image is built using the origin-ansible container. Once it has completed, we will use the latest openshift-azure build to test a cluster build. Once the image has passed a cluster build, the image is tagged as valid: true. The teardown is called and the job completes. This job uses both openshift-ansible and openshift-azure to complete its image build and deploy of a cluster.

We have e2e tests for cluster deployments in openshift-azure that use the latest published image. We don't have e2e tests for gating openshift-ansible changes, right? Or at least we need to migrate those jobs away from Jenkins. We should still be able to pull openshift-azure and run a cluster deployment in a job that runs in openshift-ansible.

**Comment by kwoodson** [comment]
@kargakis 
> We have e2e tests for cluster deployments in openshift-azure that use the latest published image. 

Correct.

> We don't have e2e tests for gating openshift-ansible changes, right? Or at least we need to migrate those jobs away from Jenkins. 

This is the beginning of that work.  The old node image jobs that are in Jenkins do not perform a cluster build.  Currently we are not testing our images which is something I was hoping to fix by merging this PR.   They used to test cluster builds but we removed that functionality when we stopped using acs-engine deployments.

> We should still be able to pull openshift-azure and run a cluster deployment in a job that runs in openshift-ansible.

This is essentially the same thing.  Whether we run the job from openshift-azure or openshift-ansible the need for each repo is equivalent.  If you prefer moving the job to openshift-ansible and using openshift-azure for images that is fine.  Currently the job is using openshift-azure's release-3.10.json and pulling the origin-ansible container image from openshift-ansible.  This felt better as we own the process for openshift-azure and we don't have to touch openshift-ansible's files/config/templates.

**Comment by kargakis** [comment]
The problem with having this in openshift-azure is additional runtime in PR
tests and extra churn to debug failures unrelated to the repo. If we want
to test node builds we should be doing that in the repo the node build code
resides (today openshift-ansible, maybe a separate repo in the future).

**Comment by kargakis** [comment]
Keeping the job definitions under openshift-azure is probably fine (we already have the Jenkins jobs there too). I see this is just a simple rewrite of the build jobs plus e2e tests so that's OK as long as we don't pull these in openshift-azure PR testing.

File: ci-operator/templates/build-azure-image-310.yaml [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+        export DEPLOY_VERSION=v${OPENSHIFT_RELEASE}
+        export RUN_SYNC_LOCAL=true
+        export IMAGE_RESOURCEGROUP=images
+
+        source /tmp/shared/environment
+        export IMAGE_RESOURCENAME=${IMAGE_NAME}
+        cd /go/src/github.com/openshift/openshift-azure/
+
+        # added timeout as this seems to loop infinitely awaiting cluster deployment
+        timeout 30m ./hack/create.sh ${RESOURCE_GROUP}-e2e
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply ci-operator/templates/build-azure-image-310.yaml]: yes, @kwoodson can you please add 45 min one on other templates too?

**Comment by petr-muller** [comment]
About the rebase request:

We have mass-renamed all ci-operator configuration files from `.json` to `.yaml` in https://github.com/openshift/release/pull/1336, because YAML is what ci-operator consumes now. All the references of these files needed to be changed too. JSON is a YAML subset so that the actual file content can stay like it is (we'll eventually mass-reformat them to the block style).

File: ci-operator/templates/build-azure-image-310.yaml [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+        ansible-playbook \
+          -e openshift_azure_resource_group_name=$RESOURCE_GROUP \
+          -e openshift_azure_input_image_ns=images \
+          -e openshift_azure_input_image_name=$IMAGE_NAME \
+          playbooks/azure/openshift-cluster/tag_image_as_valid.yml
+
+        touch /tmp/shared/exit
+
+        exit 0
```
Comment (mjudeikis) [inline ci-operator/templates/build-azure-image-310.yaml]: I think this is not needed anymore?

File: ci-operator/config/openshift/openshift-azure/release-3.10.yaml [hunk 10]
```diff
@@ -0,0 +1,65 @@
+{
+  "tag_specification": {
+    "cluster": "https://api.ci.openshift.org",
+    "namespace": "openshift",
+    "name": "origin-v3.10",
```
Comment (kargakis) [inline ci-operator/config/openshift/openshift-azure/release-3.10.yaml]: openshift-azure master is effectively origin 3.10 (we already use 3.10 images in master.yaml). I think this is the wrong place to branch node build jobs, and it's more appropriate to be using the openshift-ansible ci-operator specs.

File: ci-operator/templates/build-azure-image-310.yaml (hunk 4 above)
Comment (kargakis) [reply ci-operator/templates/build-azure-image-310.yaml]: still outstanding

File: ci-operator/templates/build-azure-image-310.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (kargakis) [reply ci-operator/templates/build-azure-image-310.yaml]: Please keep the install container the same across all templates.

File: ci-operator/templates/build-azure-image-310.yaml [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+          -e openshift_azure_install_repo=$(curl -qs https://storage.googleapis.com/origin-ci-test/releases/openshift/origin/release-${OPENSHIFT_RELEASE}/.latest-rpms) \
+          playbooks/azure/openshift-cluster/build_node_image.yml
+
+        az login --service-principal -u ${AZURE_CLIENT_ID} -p ${AZURE_CLIENT_SECRET} --tenant ${AZURE_TENANT} &>/dev/null
+        az group delete -n ${RESOURCE_GROUP} -y
+
+        touch /tmp/shared/image_create
+
+    # Runs an install
+    - name: buildcluster
```
Comment (kargakis) [inline ci-operator/templates/build-azure-image-310.yaml]: `setup` to mirror the install container in the other templates

File: ci-operator/templates/build-azure-image-310.yaml [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+
+        # added timeout as this seems to loop infinitely awaiting cluster deployment
+        timeout 30m ./hack/create.sh ${RESOURCE_GROUP}-e2e
+
+        # if completed pass generated data to other stage
+        cp -r /go/src/github.com/openshift/openshift-azure/_data /tmp/shared/
+
+        touch /tmp/shared/e2e_tests
+
+    - name: tagimage
```
Comment (kargakis) [inline ci-operator/templates/build-azure-image-310.yaml]: Is there any reason for a separate container for doing the tagging? It seems we would reduce the size of bash we maintain by moving this after running the node build playbook?

Comment (kargakis) [reply ci-operator/templates/build-azure-image-310.yaml]: Sorry, just realized you need to tag after the e2e test succeeds. Nvmd.

File: ci-operator/templates/build-azure-image-310.yaml [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+      emptyDir: {}
+    - name: shared-data
+      emptyDir: {}
+    - name: cluster-secrets-azure
+      secret:
+        secretName: azure
+
+    containers:
+    - name: buildimage
+      image: ${IMAGE_ANSIBLE}
```
Comment (kargakis) [inline ci-operator/templates/build-azure-image-310.yaml]: For ansible PRs and merged commits this needs to be `${LOCAL_IMAGE_BIN}`. For periodic jobs, it should be fine as is. Unclear how to best template that.

**Comment by kargakis** [comment]
@stevekuznetsov @bbguimaraes any suggestions on how to clone a repo (openshift-azure) in an openshift-ansible test?

**Comment by stevekuznetsov** [comment]
> @stevekuznetsov @bbguimaraes any suggestions on how to clone a repo (openshift-azure) in an openshift-ansible test?

What are you trying to acheive? This is the wrong flow.

I think this issue should have been tackled as "how does CI Operator support first-class creation and publishing of artifacts that are not container images or RPMs"

**Comment by kargakis** [comment]
The other templates we use for e2e tests in openshift-azure

**Comment by kargakis** [comment]
Apart from the lack of support for publishing non-container image related
artifacts in the ci-operator, the main issue here is that we have an
openshift-ansible playbook for building cloud images and we need to run a
cluster install on top of the image build using our own installer
(openshift-azure). Ideally, this job should gate every openshift-ansible PR
because the actual thing that is tested is the image build and the cluster
install is there for verification but we can start with a periodic.
Eventually I would love to see us move away from openshift-ansible
altogether but we haven't discussed/prioritized that yet.

**Comment by stevekuznetsov** [comment]
As a short-term hack you could use `extra_refs` in the ProwJob. Please make a card on our JIRA to track publishing non-image artifacts for repos

**Comment by kargakis** [comment]
> Please make a card on our JIRA to track publishing non-image artifacts for repos

https://jira.coreos.com/browse/DPTP-187

File: ci-operator/templates/build-azure-image.yaml [hunk 14]
```diff
@@ -0,0 +1,315 @@
+# This template is being used for testing OpenShift Ansible image builds for Azure.
```
Comment (kargakis) [inline ci-operator/templates/build-azure-image.yaml]: This template needs to be moved in `ci-operator/templates/openshift/openshift-azure`.

**Comment by kargakis** [comment]
> As a short-term hack you could use extra_refs in the ProwJob.

Is this going to work for a periodic job with `skip_cloning: true`?

File: ci-operator/templates/build-azure-image.yaml [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+        function teardown() {
+          set +e
+          mkdir -p "${HOME}"
+          export HOME=/tmp/shared
+          export DNS_DOMAIN=osadev.cloud
+          export DNS_RESOURCEGROUP=dns
+
+          cp -r /tmp/shared/_data /go/src/github.com/openshift/openshift-azure/
+          cd /go/src/github.com/openshift/openshift-azure/
+          source /etc/azure/credentials/secret
```
Comment (kargakis) [inline ci-operator/templates/build-azure-image.yaml]: You also need to source `/tmp/shared/environment` to get the resource group, right?

File: ci-operator/templates/build-azure-image.yaml [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+          -e openshift_azure_resource_group_name=$RESOURCE_GROUP \
+          -e openshift_azure_input_image_ns=images \
+          -e openshift_azure_input_image_name=$IMAGE_NAME \
+          playbooks/azure/openshift-cluster/tag_image_as_valid.yml
+
+        touch /tmp/shared/exit
+
+    # Performs cleanup of all created resources
+    - name: teardown
+      image: ${LOCAL_IMAGE_BIN}
```
Comment (kargakis) [inline ci-operator/templates/build-azure-image.yaml]: LOCAL_IMAGE_SRC should be enough I think

File: ci-operator/templates/build-azure-image.yaml [hunk 17]
```diff
@@ -0,0 +1,315 @@
...
+          playbooks/azure/openshift-cluster/build_node_image.yml
+
+        az login --service-principal -u ${AZURE_CLIENT_ID} -p ${AZURE_CLIENT_SECRET} --tenant ${AZURE_TENANT} &>/dev/null
+        az group delete -n ${RESOURCE_GROUP} -y
+
+        touch /tmp/shared/image_create
+
+    # Runs an install
+    - name: setup
+      image: ${LOCAL_IMAGE_BIN}
```
Comment (kargakis) [inline ci-operator/templates/build-azure-image.yaml]: LOCAL_IMAGE_SRC should be enough I think

File: ci-operator/templates/build-azure-image.yaml [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+        done
+        echo "Found image create marker file."
+
+        source /etc/azure/credentials/secret
+        az login --service-principal -u ${AZURE_CLIENT_ID} -p ${AZURE_CLIENT_SECRET} --tenant ${AZURE_TENANT_ID} &>/dev/null
+        # TODO: Configure these from prow
+        export DNS_DOMAIN=osadev.cloud
+        export DNS_RESOURCEGROUP=dns
+        export DEPLOY_VERSION=v${OPENSHIFT_RELEASE}
+        export RUN_SYNC_LOCAL=true
```
Comment (kargakis) [inline ci-operator/templates/build-azure-image.yaml]: remove

File: ci-operator/templates/build-azure-image.yaml [hunk 19]
```diff
@@ -0,0 +1,315 @@
...
+             echo "Exiting.  Found /tmp/shared/exit."
+             exit 1
+          fi
+          break
+        done
+        echo "Found image create marker file."
+
+        source /etc/azure/credentials/secret
+        az login --service-principal -u ${AZURE_CLIENT_ID} -p ${AZURE_CLIENT_SECRET} --tenant ${AZURE_TENANT_ID} &>/dev/null
+        # TODO: Configure these from prow
```
Comment (kargakis) [inline ci-operator/templates/build-azure-image.yaml]: remove

**Comment by kargakis** [comment]
The periodic prowjob I came up with:
```yaml
  - agent: kubernetes
    decorate: true
    name: azure-build-node-image-rhel-310
    extra_refs:
    - org: openshift
      repo: openshift-azure
      base_ref: master
    skip_cloning: true
    spec:
      containers:
      - args:
        - --secret-dir=/usr/local/e2e-azure-secret
        - --template=/usr/local/azure-image-build
        command:
        - ci-operator
        env:
        - name: CONFIG_SPEC
          valueFrom:
            configMapKeyRef:
              key: release-3.10.yaml
              # Use openshift-ansible ci-operator config for as long as
              # the node build code is part of that repo.
              name: ci-operator-openshift-openshift-ansible
        - name: JOB_NAME_SAFE
          value: azure-image-build
        - name: CLUSTER_TYPE
          value: azure
        image: ci-operator:latest
        name: test
        volumeMounts:
        - mountPath: /usr/local/azure-image-build
          name: job-definition
          subPath: azure-image-build.yaml
        - mountPath: /usr/local/e2e-azure-secret
          name: cluster-secrets-azure-file
      serviceAccountName: ci-operator
      volumes:
      - configMap:
          name: prow-job-azure-build-image
        name: job-definition
      - name: cluster-secrets-azure-file
        projected:
          sources:
          - secret:
              name: cluster-secrets-azure-file
```
Haven't actually tested but my question is, will `LOCAL_IMAGE_SRC` include the source code from `extra_refs`? @bbguimaraes @stevekuznetsov @petr-muller If so, @kwoodson I think we can get rid of the `release-3.10.yaml` ci-operator config that you are adding in this PR and use the openshift-ansible ci-operator config directly.

File: ci-operator/templates/build-azure-image.yaml (hunk 16 above)
Comment (mjudeikis) [reply ci-operator/templates/build-azure-image.yaml]: BIN and SRC one difference is that BIN has generated code and binaries built already. We use binaries for sync image build in the local namespace. Didn't checked end-to-end but I think we need bin for this, as it is used for sync image build. Just to keep consistent.

**Comment by mjudeikis** [comment]
>  I think we can get rid of the release-3.10.yaml ci-operator config that you are adding in this PR and use the openshift-ansible ci-operator config directly.

I think I know what you have in mind, but need more tech details on how it should be implemented.

File: ci-operator/templates/build-azure-image.yaml (hunk 16 above)
Comment (kargakis) [reply ci-operator/templates/build-azure-image.yaml]: But we don't care about building the sync image and we can just pull it from quay since what is tested here is the node build.

Comment (mjudeikis) [reply ci-operator/templates/build-azure-image.yaml]: our experience shows we good at this...
